diesel/src/types/mod.rs

```diff
@@ -337,18 +338,100 @@ pub enum IsNull {
     No,
 }
 
+#[derive(Debug, Clone, Copy)]
+#[doc(hidden)]
+pub struct ToSqlOutput<'a, T, DB> where
+    DB: TypeMetadata,
+    DB::MetadataLookup: 'a,
+{
+    out: T,
+    metadata_lookup: &'a DB::MetadataLookup,
+}
+
+impl<'a, T, DB: TypeMetadata> ToSqlOutput<'a, T, DB> {
+    pub fn new(out: T, metadata_lookup: &'a DB::MetadataLookup) -> Self {
+        ToSqlOutput { out, metadata_lookup }
+    }
+
+    pub fn with_buffer<U>(&self, new_out: U) -> ToSqlOutput<'a, U, DB> {
+        ToSqlOutput {
+            out: new_out,
+            metadata_lookup: self.metadata_lookup,
+        }
+    }
+
+    pub fn into_inner(self) -> T {
+        self.out
+    }
+
+    pub fn metadata_lookup(&self) -> &'a DB::MetadataLookup {
+        self.metadata_lookup
+    }
+}
+
+#[cfg(test)]
+impl<DB: TypeMetadata> ToSqlOutput<'static, Vec<u8>, DB> {
+    /// Returns a `ToSqlOutput` suitable for testing `ToSql` implementations.
+    /// Unsafe to use for testing types which perform dynamic metadata lookup.
+    pub fn test() -> Self {
+        use std::mem;
+        Self::new(Vec::new(), unsafe { mem::uninitialized() })
+    }
+}
+
+impl<'a, T: Write, DB: TypeMetadata> Write for ToSqlOutput<'a, T, DB> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
```

Comment:
There are places that it is passed to a function which expects something which implements `Write`. I could manually add `&mut *out` to those places, but the goal of this version is to avoid churning lines which don't need to change.